master/buildbot/process/buildrequestmerger.py

```diff
@@ -0,0 +1,248 @@
+import json
+import time
+
+from twisted.application import service
+from twisted.internet import reactor
+from twisted.python import log
+import sqlalchemy as sa
+
+from buildbot import config
+from buildbot.process.buildrequest import Priority
+from buildbot.util.lru import LRUCache
+
+
+class BuildRequestMerger(config.ReconfigurableServiceMixin, service.Service):
+    # Basic list of properties that must match for a buildrequest to be merged
+    # BuilderConfigs can define additional properties (see _propertiesMatch)
+    BASE_MERGE_PROPERTIES = ['force_rebuild', 'force_chain_rebuild']
+
+    def __init__(self, master):
+        self.master = master
+        self.properties_cache = LRUCache(
+            miss_fn=self.__propertiesCacheMissFn, max_size=20000)
+
+    def addBuildset(self,
+                    sourcestampsetid,
+                    reason,
+                    properties,
+                    triggeredbybrid=None,
+                    builderNames=None,
+                    external_idstring=None,
+                    _reactor=reactor):
+        """
+        Wrapper around buildsets.addBuildset that does merging before
+        buildrequests hit the db
+
+        ..seealso:: buildsets.addBuildset
+        """
+        start = time.time()
+
+        buildsetLog = {
+            'name': 'addBuildset',
+            'description': 'Log merges done while adding new buildsets',
+            'sourcestampsetid': sourcestampsetid,
+            'builderNames': builderNames,
+        }
+
+        codebase, branch, revision = self._getSourceStampInfo(sourcestampsetid)
```

Comment:
not supporting multiple codebases

Reply:
What happens when there are multiple codebases?

Will we have several entries in the `sourcestamps` table with the same `sourcestampsetid`?

Reply:
yes, that's correct